providers/dns/godaddy/godaddy.go

```diff
@@ -0,0 +1,154 @@
+	"io/ioutil"
+	"strings"
+)
+
+// GoDaddyAPIURL represents the API endpoint to call.
+const apiURL = "https://api.godaddy.com"
+
+// DNSProvider is an implementation of the acme.ChallengeProvider interface
+type DNSProvider struct {
+	apiKey    string
+	apiSecret string
+}
+
+// NewDNSProvider returns a DNSProvider instance configured for godaddy.
+// Credentials must be passed in the environment variables: GODADDY_API_KEY
+// and GODADDY_API_SECRET.
+func NewDNSProvider() (*DNSProvider, error) {
+	apikey := os.Getenv("GODADDY_API_KEY")
+	secret := os.Getenv("GODADDY_API_SECRET")
+	return NewDNSProviderCredentials(apikey, secret)
+}
+
+// NewDNSProviderCredentials uses the supplied credentials to return a
+// DNSProvider instance configured for godaddy.
+func NewDNSProviderCredentials(apiKey, apiSecret string) (*DNSProvider, error) {
+	if apiKey == "" || apiSecret == "" {
+		return nil, fmt.Errorf("GoDaddy credentials missing")
+	}
+
+	return &DNSProvider{apiKey, apiSecret}, nil
+}
+
+// Timeout returns the timeout and interval to use when checking for DNS
+// propagation. Adjusting here to cope with spikes in propagation times.
+func (c *DNSProvider) Timeout() (timeout, interval time.Duration) {
+	return 120 * time.Second, 2 * time.Second
+}
+
+func (c *DNSProvider) extractRecordName(fqdn, domain string) string {
+	name := acme.UnFqdn(fqdn)
+	if idx := strings.Index(name, "."+domain); idx != -1 {
+		return name[:idx]
+	}
+	return name
+}
+
+// Present creates a TXT record to fulfil the dns-01 challenge
+func (c *DNSProvider) Present(domain, token, keyAuth string) error {
+	fqdn, value, ttl := acme.DNS01Record(domain, keyAuth)
+	domainZone, err := c.getZone(fqdn)
+	if err != nil {
+		return err
+	}
+
+	if ttl < 600 {
+		ttl = 600
+	}
+
+	rec := DNSRecord{
+		Type: "TXT",
+		Name: c.extractRecordName(fqdn, domainZone),
+		Data: value,
+		Ttl:  ttl,
+	}
+
+	return c.updateRecord(rec, domainZone)
+}
+
+func (c *DNSProvider) updateRecord(record DNSRecord, domainZone string) error {
+	body, err := json.Marshal(record)
+	if err != nil {
+		return err
+	}
+
+	var resp *http.Response
+	resp, err = c.makeRequest("PUT", fmt.Sprintf("/v1/domains/%s/records/%s/%s", domainZone, record.Type, record.Name), bytes.NewReader(body))
+	if err != nil {
+		return err
+	}
+
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		bodyBytes, _ := ioutil.ReadAll(resp.Body)
+		return fmt.Errorf("Could not create record %v; Status: %v; Body: %s\n", string(body), resp.StatusCode, string(bodyBytes))
+	}
+	return nil
+}
+
+// CleanUp sets null value in the TXT DNS record as GoDaddy has no proper DELETE record method
+func (c *DNSProvider) CleanUp(domain, token, keyAuth string) error {
+	fqdn, _, ttl := acme.DNS01Record(domain, keyAuth)
+	domainZone, err := c.getZone(fqdn)
+	if err != nil {
+		return err
+	}
+
+	if ttl < 600 {
+		ttl = 600
+	}
+
+	rec := DNSRecord{
+		Type: "TXT",
+		Name: c.extractRecordName(fqdn, domainZone),
+		Data: "null",
```

Comment:
The problem of GoDaddy's API not having a DELETE method for records (WHY, GODADDY!?) has been documented by other GoDaddy API clients. [This python client](https://github.com/eXamadeus/godaddypy/blob/2e771c516f9b2db74f1c1d3c6721d8ef9cacd366/godaddypy/client.py#L251-L277) dealt with it by GETting all records, filtering out those of specified names and types, then PUTting them back. Since we're always dealing with specific type (TXT) and names, maybe we can PUT an empty list to `/v1/domains/{domain}/records/TXT/{name}` to clear the record, instead of changing the record to `Data: "null"`?

Reply:
I just tried my idea. Doesn't look like it will work. GoDaddy responds with an error if you PUT an empty list. I think your approach is the best way about it.

Another idea would be to `GET /v1/domains/{domain}/records/TXT`, remove the entry from the list with name `{name}`, and then PUT the reduced list back. If the list is empty, we'd have to add a dummy entry so GoDaddy will accept the request. The benefit is that there would only be at most one dummy record in there instead of a "null" record for each name. The downside is that it would be a data race if there are multiple requests at the same time.

The only downside to leaving "null" TXT records is if you consider the name of your internal subdomains to be sensitive.

Reply:
I tried it with this workaround (removing the added TXT records from the list), but it is unrealiable. Some users might have important TXT records which should never be deleted, this cannot be guaranteed by this method.

> The only downside to leaving "null" TXT records is if you consider the name of your internal subdomains to be sensitive.

A list of every certificate signed by letsencrypt is publicly available because of the transparency logs which can be searched at [crt.sh](https://crt.sh/?Identity=%25&iCAID=7395), so this is not an issue.

I think, there currently is no better way, than the implemented solution.

Reply:
Agreed. Until this is merged in, I'll be using your fork in a custom build so I can get my company's Traefik instances using Let's Encrypt. Thanks again!